crates/core/tedge_actors/src/servers/builders.rs

```diff
@@ -0,0 +1,237 @@
+    pub fn new(service_name: &str, capacity: usize) -> Self {
+        let max_concurrency = 1;
+        let (request_sender, request_receiver) = mpsc::channel(capacity);
+        let (signal_sender, signal_receiver) = mpsc::channel(4);
+        let input_receiver =
+            LoggingReceiver::new(service_name.to_string(), request_receiver, signal_receiver);
+
+        ServerMessageBoxBuilder {
+            service_name: service_name.to_string(),
+            max_concurrency,
+            request_sender,
+            input_receiver,
+            signal_sender,
+            clients: vec![],
+        }
+    }
+
+    pub fn with_max_concurrency(self, max_concurrency: usize) -> Self {
+        Self {
+            max_concurrency: std::cmp::max(1, max_concurrency),
+            ..self
+        }
+    }
+
+    /// Build a message box ready to be used by the service actor
+    fn build_service(self) -> ServerMessageBox<Request, Response> {
+        let response_sender = SenderVec::new_sender(self.clients);
+        let logging_sender = LoggingSender::new(self.service_name.clone(), response_sender);
+
+        SimpleMessageBox::new(self.input_receiver, logging_sender)
+    }
+
+    /// Build a message box aimed to concurrently serve requests
+    fn build_concurrent(self) -> ConcurrentServerMessageBox<Request, Response> {
+        let max_concurrency = self.max_concurrency;
+        let clients = self.build_service();
+        ConcurrentServerMessageBox::new(max_concurrency, clients)
+    }
+}
+
+impl<Req: Message, Res: Message> RuntimeRequestSink for ServerMessageBoxBuilder<Req, Res> {
+    fn get_signal_sender(&self) -> DynSender<RuntimeRequest> {
+        self.signal_sender.sender_clone()
+    }
+}
+
+impl<Req: Message, Res: Message> ServiceProvider<Req, Res, NoConfig>
+    for ServerMessageBoxBuilder<Req, Res>
+{
+    fn add_peer(&mut self, peer: &mut impl ServiceConsumer<Req, Res, NoConfig>) {
+        let client_id = self.clients.len();
+        let request_sender = KeyedSender::new_sender(client_id, self.request_sender.clone());
+
+        self.clients.push(peer.get_response_sender());
+        peer.set_request_sender(request_sender)
+    }
+}
+
+impl<Req: Message, Res: Message> Builder<ServerMessageBox<Req, Res>>
+    for ServerMessageBoxBuilder<Req, Res>
+{
+    type Error = Infallible;
+
+    fn try_build(self) -> Result<ServerMessageBox<Req, Res>, Self::Error> {
+        Ok(self.build())
+    }
+
+    fn build(self) -> ServerMessageBox<Req, Res> {
+        self.build_service()
+    }
+}
+
+impl<Req: Message, Res: Message> Builder<ConcurrentServerMessageBox<Req, Res>>
+    for ServerMessageBoxBuilder<Req, Res>
+{
+    type Error = Infallible;
+
+    fn try_build(self) -> Result<ConcurrentServerMessageBox<Req, Res>, Self::Error> {
+        Ok(self.build())
+    }
+
+    fn build(self) -> ConcurrentServerMessageBox<Req, Res> {
+        self.build_concurrent()
+    }
+}
+
+pub struct Concurrent;
+pub struct Sequential;
```

Comment:
`max_concurrency = 1` means sequential? Then, why do we need to have this?

Reply:
Yes, there is something to improve here.